### PR TITLE
New bootstrap

### PR DESCRIPTION
**Pull request by dra27**
OK, here goes!

The good:
- There are now no .boot.ml files 🎉
- It's faster on everything I've tried it on for `-j 1`
- It's comparable and usually faster for `-j 4`
- It integrates my previous work on the secondary compiler, so this dune can be installed in a switch running OCaml 4.02+:

```bash
opam switch create dune-secondary --repos=secondary=git+https://github.com/dra27/opam-repository#secondary ocaml-base-compiler.4.02.3
opam install ocamlfind-secondary
```

The bad:
- The scheduler is not as good on high performance machines - on our 72 core machines, this bootstrap's taking 9sec or so where the old one took 6sec.
- The scheduler seems to suck on Windows (it's actually faster not to parallelise) - this isn't true of Dune itself (although the `-j 1` build on Windows of this bootstrap is comparable to a parallel build on Windows). Not sure whether it's the scheduler, Windows or both. Some attention should at some point be put to the fact that a 32-core high performance Windows VM takes six times as long to compile and link dune as 4-core Linux VM on my laptop...

The still-to-do:
- Work out suitable parallelism
- `setup.ml` is written to the tree at the moment (i.e. `configure.ml` support needs to go back in)
- bootstrap info needs to be fed into the script properly
- I have no story at the moment for the build_info

I've tried various hybrid schemes on the way to this - apologies in advance for my generally low standard of code.

### CODE REVIEW

**Review by ghost — DISMISSED** [review]

File: .travis.yml [hunk 1]
```diff
@@ -15,10 +15,7 @@ script: bash -ex .travis-ci.sh build
 matrix:
   include:
   - os: linux
-    env: OCAML_VERSION=4.06 OCAML_RELEASE=1 WITH_OPAM=0
-    stage: Build
-  - os: linux
-    env: OCAML_VERSION=4.07 OCAML_RELEASE=1 WITH_OPAM=0
+    env: OCAML_VERSION=4.02 OCAML_RELEASE=3 WITH_OPAM=0
```
[the remark this answers is not in the document]
Comment (dra27) [reply .travis.yml]: The bootstrap requires 4.08, falling back to ocaml-secondary - do I need to make sure that it can work with 4.06 too?

**Comment by diml** [comment]
I did the build info thing

File: boot/duneboot.ml [hunk 2]
```diff
@@ -664,7 +664,7 @@ let scheduler parallel_count name exe external_includes external_libraries
           let task =
             let cwd = Sys.getcwd () in
             Sys.chdir build_dir;
-            Printf.printf "%s -c -g -no-alias-deps -w -49 -I +threads %s\n"
+            Printf.printf "%s -c -g -no-alias-deps -w -49 -I +threads %s\n%!"
```
Comment (dra27) [inline boot/duneboot.ml]: I should have put a COMBAK on this - the flush considerably slows the build down (same with `dune` itself)

Comment (dra27) [reply boot/duneboot.ml]: (I've also left things hard-coded in the string, but that's another story...)

Comment (diml) [reply boot/duneboot.ml]: Ok. I suppose we can do the same as for Dune then: we only print command that fail

Comment (diml) [reply boot/duneboot.ml]: Or we can also print nothing. The output of the compiler already contains `File ...`

Comment (dra27) [reply boot/duneboot.ml]: On my Ubuntu 19.04 4-core VirtualBox (so at the "console"): 

```
dra@disco:~/dune-vanilla$ time make
…
real	0m36.229s
user	1m5.748s
sys	0m23.141s

dra@disco:~/dune-vanilla$ git clean -dfx &>/dev/null
dra@disco:~/dune-vanilla$ time make &>/dev/null

real	0m21.695s
user	0m42.232s
sys	0m11.225s
```

On our EPYCs (via SSH):

```
dra27@bremusa:~/dune$ time make
…
real    0m10.891s
user    0m34.895s
sys     0m15.863s

dra27@bremusa:~/dune$ git clean -dfx &>/dev/null
dra27@bremusa:~/dune$ time make &>/dev/null

real    0m10.984s
user    0m34.487s
sys     0m16.254s
```

but on its console, without piping to `/dev/null`:

```
real    0m14.198s
user    0m35.433s
sys     0m16.273s
```

I think that when `stdout` is a tty, `dune` should be more conservative about flushing.

Comment (diml) [reply boot/duneboot.ml]: I just hid the compilation commands. We could just dump them into a log file.

File: boot/duneboot.ml [hunk 3]
```diff
@@ -0,0 +1,1189 @@
...
+  if Sys.file_exists b then fatal "%s already exists" b;
+  let ic = open_in_bin a in
+  let len = in_channel_length ic in
+  let s = really_input_string ic len in
+  close_in ic;
+  let oc = open_out_bin b in
+  output_string oc header;
+  (* MSVC is fussy about #line 1 vs # 1 *)
+  let directive = if Filename.extension b = ".c" then "line" else "" in
+  fprintf oc "#%s 1 %S\n" directive a;
```
Comment (dra27) [inline boot/duneboot.ml]: Happy with this fix, @diml?

Comment (diml) [reply boot/duneboot.ml]: Sure!

**Pull request merged**